internal/controller/snapshot/snapshot_adapter.go

```diff
@@ -1058,3 +1072,68 @@ func (a *Adapter) findSnapshotWithOpenedPR(snapshot *applicationapiv1alpha1.Snap
+// removes finalizer before the pipelinerun is set as CancelledRunFinally to be gracefully cancelled
+func (a *Adapter) checkAndCancelOldSnapshotsPipelineRun(application *applicationapiv1alpha1.Application, snapshot *applicationapiv1alpha1.Snapshot) error {
+
+	snapshots, err := a.loader.GetAllSnapshotsForPR(a.context, a.client, application, snapshot.GetLabels()[gitops.PipelineAsCodePullRequestAnnotation], snapshot.GetLabels()[gitops.PipelineAsCodeRepoURLAnnotation])
+	if err != nil {
+		a.logger.Error(err, "Failed to fetch Snapshots for the application",
+			"application.Name:", application.Name)
+		return err
+	}
+	sortedSnapshots := gitops.SortSnapshots(*snapshots)
+	if len(sortedSnapshots) < 2 {
+		a.logger.Info("There are no snapshots found for given repo at the moment.")
+		return nil
+	}
+
+	for i := 1; i < len(sortedSnapshots); i++ {
+		if gitops.IsComponentSnapshot(&sortedSnapshots[i]) {
+			err = gitops.MarkSnapshotAsCanceled(a.context, a.client, &sortedSnapshots[i], "Snapshot canceled/superseded")
+			if err != nil {
+				a.logger.Error(err, "Failed to mark snapshot as canceled", "snapshot.Name", &sortedSnapshots[i].Name)
+				return err
+			}
+			err = a.cancelAllPipelineRunsForSnapshot(&sortedSnapshots[i])
+		}
+	}
+	return err
+}
+
+// cancelAllPipelineRunsForSnapshot gets all integration test pipelieruns for a given snapshot
+// sorts the pipelineruns according to the time they were created and cancels all but the latest pipelineruns
+func (a *Adapter) cancelAllPipelineRunsForSnapshot(snapshot *applicationapiv1alpha1.Snapshot) error {
+	// get all integration pipelineruns for a snapshot
+	integrationTestPipelineruns, err := a.loader.GetAllIntegrationPipelineRunsForSnapshot(a.context, a.client, snapshot)
+	if err != nil {
+		a.logger.Error(err, "Failed to get all integration pipelineruns for snapshot", "snapshot.Name", snapshot.Name)
+		return err
+	}
+
+	for _, plr := range integrationTestPipelineruns {
+		plr := plr
+		pipelinerunStatus, _, err := h.GetRunningIntegrationPipelineRunStatus(&plr)
+		if err != nil {
+			a.logger.Error(err, "Failed to get integration test pipelinerun status", "snapshot.Name", plr.Name)
+			return err
+		}
+		if pipelinerunStatus == intgteststat.IntegrationTestStatusInProgress {
+			//remove finalizer and cancel pipelinerun
+			err = h.RemoveFinalizerFromPipelineRun(a.context, a.client, a.logger, &plr, h.IntegrationPipelineRunFinalizer)
+			if err != nil {
+				a.logger.Error(err, "Failed to remove finalizer from integration test pipelinerun")
+				return err
+			}
+			// set "CancelledRunFinally" to PLR status, should gracefully cancel pipelinerun, this is so raw I hate this
+			patch := client.MergeFrom(plr.DeepCopy())
+			plr.Spec.Status = tektonv1.PipelineRunSpecStatusCancelledRunFinally
+			err := a.client.Patch(a.context, &plr, patch)
+			if err != nil {
+				return err
+			}
```

Comment:
nit: since we're updating(patching) the pipelineRun here, it might be useful to add an audit log call here, similar to how we emit one when [creating the pipelineRun](https://github.com/konflux-ci/integration-service/blob/c23fc034faec01744e823881285f265196f16a4b/internal/controller/snapshot/snapshot_adapter.go#L777-L778).
That said, we already emit an audit log when [removing the finalizer](https://github.com/konflux-ci/integration-service/blob/c23fc034faec01744e823881285f265196f16a4b/helpers/integration.go#L433), so I'd also be fine with not adding another one.